lib/Dialect/FIRRTL/Transforms/InferDomains.cpp

```diff
@@ -0,0 +1,1426 @@
+      auto domainOp = globals.circuitInfo.getDomain(i);
+      if (!first) {
+        out << ", ";
+        first = false;
+      }
+      out << domainOp.getName() << ": ";
+      render(out, idTable, row->elements[i]);
+    }
+    out << "]";
+    return;
+  }
+}
+
+template <typename T>
+void InferModuleDomains::emitPortDomainCrossingError(T op, size_t i,
+                                                     size_t domainTypeID,
+                                                     Term *term1,
+                                                     Term *term2) const {
+  VariableIDTable idTable;
+
+  auto portName = op.getPortNameAttr(i);
+  auto portLoc = op.getPortLocation(i);
+  auto domainDecl = globals.circuitInfo.getDomain(domainTypeID);
+  auto domainName = domainDecl.getNameAttr();
+
+  auto diag = emitError(portLoc);
+  diag << "illegal " << domainName << " crossing in port " << portName;
+
+  auto &note1 = diag.attachNote();
+  note1 << "1st instance: ";
+  render(note1, idTable, term1);
+
+  auto &note2 = diag.attachNote();
+  note2 << "2nd instance: ";
+  render(note2, idTable, term2);
+}
+
+template <typename T>
+void InferModuleDomains::emitDomainPortInferenceError(T op, size_t i) const {
+  auto name = op.getPortNameAttr(i);
+  auto diag = emitError(op->getLoc());
+  auto info = op.getDomainInfo();
+  diag << "unable to infer value for domain port " << name;
+  for (size_t j = 0, e = op.getNumPorts(); j < e; ++j) {
+    if (auto assocs = dyn_cast<ArrayAttr>(info[j])) {
+      for (auto assoc : assocs) {
+        if (i == cast<IntegerAttr>(assoc).getValue()) {
+          auto name = op.getPortNameAttr(j);
+          auto loc = op.getPortLocation(j);
+          diag.attachNote(loc) << "associated with hardware port " << name;
+          break;
+        }
+      }
+    }
+  }
+}
+
+static LogicalResult inferModuleDomains(GlobalState &globals,
+                                        FModuleOp module) {
+  return InferModuleDomains::run(globals, module);
+}
+
+//===---------------------------------------------------------------------------
+// InferDomainsPass: Top-level pass implementation.
+//===---------------------------------------------------------------------------
+
+static LogicalResult runOnModuleLike(InferDomainsMode mode,
+                                     GlobalState &globals, Operation *op) {
+  if (auto module = dyn_cast<FModuleOp>(op)) {
+    if (shouldInfer(module, mode))
+      return inferModuleDomains(globals, module);
+    return checkModuleDomains(globals, module);
+  }
+
+  if (auto extModule = dyn_cast<FExtModuleOp>(op)) {
+    return checkModuleDomains(globals, extModule);
+  }
+
+  return success();
+}
+
+namespace {
+struct InferDomainsPass
+    : public circt::firrtl::impl::InferDomainsBase<InferDomainsPass> {
+  using InferDomainsBase::InferDomainsBase;
+  void runOnOperation() override;
+};
+} // namespace
+
+void InferDomainsPass::runOnOperation() {
+  LLVM_DEBUG(debugPassHeader(this) << "\n");
```

Comment:
After a suggestion from @dtzSiFive, there is now a better RAII utility for doing this. Try:


```suggestion
  CIRCT_DEBUG_SCOPED_PASS_LOGGER(this);
```

This can then drop the footer (as the footer is created on destruction of the underlying object).